File: findDiagonalOrder.java
//TC: O(mxn) as we traverse through the elements in matrix
//SP\C: O(n) as we are storing n elements
//Leetcode: Success!
//Problems: switching from Python to Java

class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        //Sanity check
        if(matrix==null || matrix.length==0) return new int[0];
        
        // Initialize the rows and cols length
        int m = matrix.length;
        int n = matrix[0].length;
        
        //Initialize resultant matrix
        int[] res = new int[m*n];
        
        int i = 0; int r = 0; int c = 0; int dir = 1;
        while(i<m*n){
            // initialize resultant array
            
            res[i] = matrix[r][c];
            // Upward direction
            if(dir==1){
                // check if the at the last column
                if(c==n-1){
                    r++;
                    dir = -1;
                }
                // check if at first row
                else if(r==0){
                    c++;
                    dir = -1;
                }
                else{
                    r--;
                    c++;
                }
            }
            // Downward direction
            else{
                // check if the at the last row
                if(r==m-1){
                    c++;
                    dir = 1;
                }
                // check if at first column
                else if(c==0){
                    r++;
                    dir = 1;
                }
                else{
                    r++;
                    c--;
                }
            }
            i++;
        }
        return res;
    }
}

File: findDiagonalOrder.py
# Promlems with the 2D matrix. How do I aasign vlaues from 2D matrix to list without going out of bounds!!!?!!?

import numpy as np
class Solution(object):
    def findDiagonalOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        # length of the matrix for number of rows
        m = len(matrix)
        # length of the matrix for number of cols
        n = len(matrix[0])
        # initializing the variables for counters
        i = 0
        r = 0
        c = 0
        dir = 1 # using direction as flag
        resultant = []*(m*n)
        print(m*n)
        while(i<m*n):
            print(matrix[r][c])
            x = matrix[r][c]
            resultant[i].append(x)
            print(resultant[i])
            # to move in the upward direction
            if dir==1:
                if c==n-1: # if reached the last column
                    r+=1
                    dir = -1
                elif r==0: # if at the first row
                    c+=1
                    dir = -1
                else: # normally loop increment the variables
                    r-=1
                    c+=1
            else:
                if r==m-1:
                    c+=1
                    dir = 1
                elif c==0:
                    r+=1
                    dir = 1
                else:
                    r+=1
                    c-=1
            i +=1
        return resultant

s = Solution()
matrix = [
    [ 1, 2, 3 ],
    [ 4, 5, 6 ],
    [ 7, 8, 9 ]
          ]
res = s.findDiagonalOrder(matrix)
print(res)
